File: src/org/warp/picalculator/math/rules/UndefinedRule2.java
package org.warp.picalculator.math.rules;

import java.util.ArrayList;

import org.warp.picalculator.Error;
import org.warp.picalculator.math.Calculator;
import org.warp.picalculator.math.functions.Division;
import org.warp.picalculator.math.functions.Function;
import org.warp.picalculator.math.functions.Number;
import org.warp.picalculator.math.functions.Undefined;

/**
 * Undefined rule<br>
 * <b>a / 0 = undefined</b>
 * @author Andrea Cavalli
 *
 */
public class UndefinedRule2 {

	public static boolean compare(Function f) {
		Calculator root = f.getRoot();
		Division fnc = (Division) f;
		if (fnc.getVariable2() instanceof Number) {
			Number numb = (Number) fnc.getVariable2();
			if (numb.equals(new Number(root, 0))) {
				return true;
			}
		}
		return false;
	}

	public static ArrayList<Function> execute(Function f) throws Error {
		Calculator root = f.getRoot();
		ArrayList<Function> result = new ArrayList<>();
		result.add(new Undefined(root));
		return result;
	}

}

File: src/main/resources/rules/UndefinedRule2.js
//Imports


var Error = org.warp.picalculator.Error;
var Function = org.warp.picalculator.math.Function;
var MathContext = org.warp.picalculator.math.MathContext;
var Division = org.warp.picalculator.math.functions.Division;
var Number = org.warp.picalculator.math.functions.Number;
var Undefined = org.warp.picalculator.math.functions.Undefined;

var ObjectArrayList = Java.type("it.unimi.dsi.fastutil.objects.ObjectArrayList");

/**
 * Undefined rule
 * a / 0 = undefined
 * 
 * @author Andrea Cavalli
 *
 */
var rule = {
	// Rule name
	getRuleName: function() {
		return "UndefinedRule2";
	},
	// Rule type
	getRuleType: function() {
		return RuleType.EXISTENCE;
	},
	/* Rule function
	   Returns:
	     - null if it's not executable on the function "f"
	     - An ObjectArrayList<Function> if it did something
	*/

	execute: function(f) {
		var isExecutable;
		if (ScriptUtils.instanceOf(f, Division.class)) {
			var root = f.getMathContext();
			var fnc = f;
			if (ScriptUtils.instanceOf(fnc.getParameter2(), Number.class)) {
				var numb = fnc.getParameter2();
				if (numb.equals(new Number(root, 0))) {
					isExecutable = true;
				}
			}
		}
	
		if (isExecutable) {
			var root = f.getMathContext();
			var result = new ObjectArrayList();
			result.add(new Undefined(root));
			return result;
		} else {
			return null;
		}
	}
}

//Add this rule to the list of rules
RulesManager.addRule(engine.getInterface(rule, Rule.class));